### PR TITLE
Refactors pipeline providers to support IoC and dependency injection

### PR DESCRIPTION
**Pull request by wbreza**
- Moves pipeline related files from `pkg/commands/pipeline` to `pkg/pipeline` to be consistent with other packages
- Updates `pipeline.Manager` and supporting CI/SCM providers to support IoC/DI
- Updates `azd pipeline config` to simplify dependencies

Supports #2356

### CODE REVIEW

File: cli/azd/pkg/pipeline/pipeline_manager.go [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+}
+
+type PipelineConfigResult struct {
+	RepositoryLink string
+	PipelineLink   string
+}
+
+// PipelineManager takes care of setting up the scm and pipeline.
+// The manager allows to use and test scm providers without a cobra command.
+type PipelineManager struct {
```
Comment (wbreza) [inline cli/azd/pkg/pipeline/pipeline_manager.go:57]: Moved file.

File: cli/azd/pkg/pipeline/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package pipeline
```
Comment (wbreza) [inline cli/azd/pkg/pipeline/pipeline.go:4]: Moved file.

File: cli/azd/cmd/container.go [hunk 3]
```diff
@@ -367,6 +368,25 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 		}
 	}
 
+	// Pipelines
+	container.RegisterSingleton(pipeline.NewPipelineManager)
```
Comment (vhvb1989) [inline cli/azd/cmd/container.go:372]: I am a little nervous about what the container will be resolving to initialize the manager.
Unfortunately, we need to run manual tests for this, as the unit-tests for picking the appropriate provider mocks the input env and args.

Manual tests we need:
- github as default provider when no cmd-args and no azure.yaml config
- github when no cmd-args and github set in azure.yml
- azdo when no cmd-args and azdo set in azure.yml
- github when cmd-args=github (when there is azure.yml config and when there isn't)
- azdo  when cmd-args=azdo (when there is azure.yml config and when there isn't)
- azdo when there is NO cmd-args and no azure.yml config, but previous run used cmd-args=azdo (selection available in .env)

I can find sometime in the future to run these tests manually (if you haven't).  That's the only thing I think could break out of the refactoring,  if the container is not resolving or reading config and/or env at the right time

**Review by vhvb1989 — APPROVED** [review]
Added a comment about some manual tests we need before merging.
Other than that, LGTM!!  
Thank you for doing this.

File: cli/azd/cmd/container.go (hunk 3 above)
Comment (wbreza) [reply cli/azd/cmd/container.go:372]: These are the same test cases that are in the pipeline manager unit tests.  The configuration from env, project config, CLI params all come from the container.

I also manually ran through all these test cases and they are working as expected.

**Review by ellismg — APPROVED** [review]
Wonderful cleanup!  Thank you for moving more stuff on to the DI plan.  This should make future work I want to do in this area related to the intersection of resource group scoped deployments and `pipeline config` so much more nicer!

**Pull request merged**